rfd/XXXX-upgrade-system.md

```diff
@@ -0,0 +1,659 @@
+installed versions.
+
+In this maximal usecase, teleport is responsible for discovery, selection, coordination, validation, and
+monitoring.  Most importantly, teleport must do all of this in a secure and reliable manner.  The potential
+fallout from bugs and vulnerabilities is greater than in the minimal usecase.
+
+#### Hacker Usecase
+
+Cluster administrators march to the beat of their own drum.  They want to know the latest publically available
+teleport releases, but skip all prime number patches.  Nodes can only be upgraded if it is low
+tide in their region, the moon is waxing, and the ISS is on the other side of the planet. They want to use
+teleport's native download and verification logic, but they also need to start the downloaded binary in
+a sandbox first to ensure it won't trigger their server's self-destruct.  If rollback is necessary, the rollback
+reason and timestamp need to be steganographically encoded into a picture of a turtle and posted to instagram.
+
+This usecase has essentially the same requirements as the maximual usecase, with one addition.  It necessitates
+*loose coupling* of componentes.
+
+
+### Security
+
+#### General
+
+- The `version-directive` must specify concrete version numbers, and provide any metadata necessary for
+package validation (varies by install method).  Any automated system which generates `version-directive`s
+(e.g. a TUF `version-controller` or some external version control plugin) must not fail silently (may indicate a censorship attack).
+Teleport can assist with this by emitting warnings if the directive is outdated.
+
+- Teleport should detect and warn about invalid state-transitions proposed by a `version-directive`, and
+refuse to enact them (e.g. refusing to downgrade from `v3.4.5` to `v1.2.3`). These checks should be independent
+of a specific controller (i.e. they should apply to custom plugins too).
+
+- We should encourage decentralized trust. The TUF based system should leverage TUF's multisignature support
+to ensure that compromise of a single key cannot compromise installations. We should also provide tools
+to help those using custom installation mechanism to avoid single-point failures as well (e.g. mutliparty approval 
+for pending `version-directive`s).
+
+- TODO
+
+
+#### TUF-Specific Security Model
+
+- TODO: explain TUF security model, thresholded signing + deterministic compilation, `teleport verify`,
+fips compat, etc.
+
+
+### Inventory Status and Control Model
+
+- Auth servers exert direct control over non-auth instance upgrades via bidirectional
+GRPC control stream.
+
+- Non-auth instances advertise detailed information about the current isntallation, and
+implement handlers for control stream messages that can execute whatever local component
+is required for running a given install method (e.g. executing a specific script if the
+`local-script` installer is in use).
+
+- Each control stream is registered with a single auth server, so each auth server is
```

Comment:
We can steal the idea of rollingUpgradePolicy from K8s with a simple upgrade policy tracking the failure threshold and stopping at it.